### PR TITLE
Make `RelationValue` hashable.

### PR DESCRIPTION
**Pull request by sallner**
At the moment the C-Implementation does not seem to support hash in Python 3.

We need this in https://github.com/plone/plone.app.linkintegrity/blob/master/plone/app/linkintegrity/handlers.py#L98 as we want to have a set.

I have no idea, why travis is not running the tests. Is it somehow deactivated?

### CODE REVIEW

File: src/z3c/relationfield/relation.py [hunk 1]
```diff
@@ -21,6 +21,10 @@ def __init__(self, to_id):
         self.__parent__ = None
         self.from_attribute = None
 
+    def __hash__(self):
+        """There should only be one RelationVaule per sort_key."""
+        return id(self)
```
Comment (icemac) [inline src/z3c/relationfield/relation.py]: According to https://stackoverflow.com/questions/17192418/hash-function-in-python#17192858 the hash function is implemented as `id(obj) // 16`. Maybe we should use this here, too.

**Review by icemac — COMMENTED** [review]
I cannot tell whether this is the right approach.

**Comment by icemac** [comment]
> I have no idea, why travis is not running the tests. Is it somehow deactivated?

I reinstalled the TravisCI service in the repository and detected that GitHub services are going to get deprecated. I additionally installed the Travis App from GitHub Marketplace. Now one of them has triggered. (Previously Travis did not even know about the branch and the PR.)

**Comment by mgedmin** [comment]
It looks like a dependency, z3c.objpath, is not Python 3 compatible?

(Looks like Py3 compat was merged to z3c.objpath's git master, but nobody made a 1.2 -- or 2.0? -- release to PyPI.)

**Comment by ale-rt** [comment]
z3c.objpath might need a release

**Comment by icemac** [comment]
See zopefoundation/z3c.objpath#3 with the request for a release.

**Comment by icemac** [comment]
z3c.objpath is now now released, restarting the Travis jobs.

**Comment by ale-rt** [comment]
I am working on making this PR pass on Python 3.
We need to:
1. fix master (see #13)
2. implement ordering
3. change zope.copy to use the latest protocol when pickling/unpickling otherwise this error will occur:
```
Error in test /home/ale/Code/plone/packages/z3c.relationfield/src/z3c/relationfield/README.rst
Traceback (most recent call last):
  File "/usr/lib/python3.6/doctest.py", line 2282, in debug
    runner.run(self._dt_test, clear_globs=False)
  File "/usr/lib/python3.6/doctest.py", line 1839, in run
    r = DocTestRunner.run(self, test, compileflags, out, False)
  File "/usr/lib/python3.6/doctest.py", line 1476, in run
    return self.__run(test, compileflags, out)
  File "/usr/lib/python3.6/doctest.py", line 1382, in __run
    exception)
  File "/usr/lib/python3.6/doctest.py", line 1845, in report_unexpected_exception
    raise UnexpectedException(test, example, exc_info)
doctest.UnexpectedException: <DocTest README.rst from /home/ale/Code/plone/packages/z3c.relationfield/src/z3c/relationfield/README.rst:0 (186 examples)>

AttributeError: Can't get attribute 'Item' on <module 'builtins' (built-in)>

> /home/ale/.buildout/eggs/zope.copy-4.1.0-py3.6.egg/zope/copy/__init__.py(37)clone()
```

**Comment by ale-rt** [comment]
If we could have this merged and released https://github.com/zopefoundation/zope.copy/pull/4 we could fix the build merging #13 and adding this patch:
```diff
diff --git a/src/z3c/relationfield/README.rst b/src/z3c/relationfield/README.rst
index 103c0bf..3fef866 100644
--- a/src/z3c/relationfield/README.rst
+++ b/src/z3c/relationfield/README.rst
@@ -306,7 +306,7 @@ We can also sort relations:
 
 .. code-block:: python
 
-  >>> expected = [('', u'a'), ('', u'b'), (u'b', u'a')]
+  >>> expected = [(u'b', u'a'), ('', u'a'), ('', u'b')]
   >>> observed = [(rel.from_path, rel.to_path) for rel in
   ...  sorted([root['b'].rel, rel_to_a, rel_to_b])]
   >>> expected == observed
diff --git a/src/z3c/relationfield/relation.py b/src/z3c/relationfield/relation.py
index fd6cd76..2d0a9d0 100644
--- a/src/z3c/relationfield/relation.py
+++ b/src/z3c/relationfield/relation.py
@@ -1,3 +1,4 @@
+from functools import total_ordering
 from persistent import Persistent
 from z3c.objpath.interfaces import IObjectPath
 from z3c.relationfield.interfaces import IRelationValue
@@ -10,6 +11,7 @@ from zope.intid.interfaces import IIntIds
 
 
 @implementer(IRelationValue)
+@total_ordering
 class RelationValue(Persistent):
 
     _broken_to_path = None
@@ -23,7 +25,13 @@ class RelationValue(Persistent):
 
     def __hash__(self):
         """There should only be one RelationVaule per sort_key."""
-        return id(self)
+        return id(self) // 16
+
+    def __eq__(self, other):
+        return self.__hash__() == other.__hash__()
+
+    def __le__(self, other):
+        return self.__hash__() <= other.__hash__()
 
     @property
     def from_id(self):
```

**Comment by ale-rt** [comment]
This should be ready for another review @icemac, @mgedmin

File: src/z3c/relationfield/relation.py [hunk 2]
```diff
@@ -73,10 +79,18 @@ def __eq__(self, other):
...
+        if (self.from_attribute or '') < (other.from_attribute or ''):
+            return True
+        if (self.from_path or '') < (other.from_path or ''):
+            return True
+        if (self.to_path or '') < (other.to_path or ''):
+            return True
+        return False
 
     def _sort_key(self):
         return (self.from_attribute, self.from_path, self.to_path)
```
Comment (mgedmin) [inline src/z3c/relationfield/relation.py:96]: If from_attribute/from_path/to_path can be None, won't this cause TypeErrors on Python 3?

File: src/z3c/relationfield/relation.py [hunk 3]
```diff
@@ -21,6 +23,10 @@ def __init__(self, to_id):
         self.__parent__ = None
         self.from_attribute = None
 
+    def __hash__(self):
+        """There should only be one RelationVaule per sort_key."""
+        return id(self) // 16
```
Comment (mgedmin) [inline src/z3c/relationfield/relation.py:28]: I'm not sure it's safe to have persistent objects whose hashes change every time they're loaded from the ZODB.  I'd have to think about this more.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+envlist =
+   py27,
+   py36,
+   py37
+
+[testenv]
+commands =
+    {envbindir}/buildout -c {toxinidir}/buildout.cfg buildout:directory={envdir} buildout:develop={toxinidir} bootstrap
+    {envbindir}/buildout -c {toxinidir}/buildout.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
+    {envbindir}/test
```
Comment (mgedmin) [inline tox.ini:11]: (I'm also not too happy about buildout being used inside tox.  Other packages tend to call zope-testrunner directly)

File: src/z3c/relationfield/relation.py [hunk 5]
```diff
@@ -73,10 +79,18 @@ def __eq__(self, other):
...
+        the path of the object the relation is one and the path of the object
+        the relation is pointing to.
+        '''
+        if (self.from_attribute or '') < (other.from_attribute or ''):
+            return True
+        if (self.from_path or '') < (other.from_path or ''):
+            return True
+        if (self.to_path or '') < (other.to_path or ''):
+            return True
+        return False
```
Comment (mgedmin) [inline src/z3c/relationfield/relation.py:93]: Could this be simplified to
```python
return self._sort_key() < other._sort_key()
```
?

And also maybe we should check object types, i.e.

```python
if not isinstance(other, RelationValue):
    return NotImplemented
```

File: src/z3c/relationfield/relation.py (hunk 3 above)
Comment (mgedmin) [reply src/z3c/relationfield/relation.py:28]: Also, `id()` is very PyPy-unfriendly.  Maybe use `return object.__hash__(self)`?

Comment (ale-rt) [reply src/z3c/relationfield/relation.py:28]: Fine for me!

File: src/z3c/relationfield/relation.py (hunk 2 above)
Comment (ale-rt) [reply src/z3c/relationfield/relation.py:96]: That's why I am adding `or ''`, to always compare str

File: src/z3c/relationfield/relation.py (hunk 5 above)
Comment (ale-rt) [reply src/z3c/relationfield/relation.py:93]: Checking the type is a nice plus.
comparing `obj._sort_key()` will return errors when the tuple contains `None` and we need to check for `None` in the `__eq__` method

File: src/z3c/relationfield/relation.py (hunk 2 above)
Comment (mgedmin) [reply src/z3c/relationfield/relation.py:96]: But not in `_sort_key()`.

If `_sort_key()` is no longer used, it should be removed.  If it is still used, it should handle sorting without TypeError when some of the attributes are None.

File: src/z3c/relationfield/relation.py (hunk 5 above)
Comment (mgedmin) [reply src/z3c/relationfield/relation.py:93]: Oh, there's an `__eq__`?

Do we preserve the requirement that hash(x) == hash(y) when `x.__eq__(y)` returns True?  If not, that's bad.

**Comment by icemac** [comment]
@esteele Hm, I'd like to have seen the issues in this PR resolved and merged before using it as a base for a release.

**Comment by esteele** [comment]
Damn. I didn't realize we weren't on master. I'm sorry about that.

**Comment by esteele** [comment]
@icemac What's the best way to go about making this right?

**Comment by icemac** [comment]
@esteele I'd suggest to get the PR ready to merge before making any other commits to the `master` branch.
After it is merged, it does not matter that the release was cut from this branch instead of `master`.

**Comment by esteele** [comment]
Ok. I'll see if we can get this wrapped up during the sprints this weekend.

**Comment by mgedmin** [comment]
I'm tempted to suggest just merging this, given that it's been released to PyPI.

Any refactorings/improvements from this point onwards need to consider backwards compatibility against the code that was released.

**Comment by icemac** [comment]
@mgedmin You are right. Merging this PR is the best step forward.

**Comment by icemac** [comment]
Thank you all!

**Pull request merged**